File: aether_flow_flutter/lib/shared_input_event.dart
import 'dart:convert';

/// 共享輸入事件類型
enum InputEventType {
  mouseMove,
  mouseLeftDown,
  mouseLeftUp,
  mouseRightDown,
  mouseRightUp,
  keyDown,
  keyUp,
  scroll,
}

/// 共享輸入事件 (滑鼠/鍵盤 - Flutter 實作)
class SharedInputEvent {
  final InputEventType type;
  final double? x; // 標準化座標 (0.0 - 1.0)
  final double? y;
  final int? keyCode;
  final double? deltaX;
  final double? deltaY;
  final int timestamp;

  SharedInputEvent({
    required this.type,
    this.x,
    this.y,
    this.keyCode,
    this.deltaX,
    this.deltaY,
    int? timestamp,
  }) : timestamp = timestamp ?? DateTime.now().millisecondsSinceEpoch;

  Map<String, dynamic> toJson() {
    return {
      'type': type.index,
      'x': x,
      'y': y,
      'keyCode': keyCode,
      'deltaX': deltaX,
      'deltaY': deltaY,
      'timestamp': timestamp,
    };
  }

  factory SharedInputEvent.fromJson(Map<String, dynamic> json) {
    return SharedInputEvent(
      type: InputEventType.values[json['type'] as int],
      x: (json['x'] as num?)?.toDouble(),
      y: (json['y'] as num?)?.toDouble(),
      keyCode: json['keyCode'] as int?,
      deltaX: (json['deltaX'] as num?)?.toDouble(),
      deltaY: (json['deltaY'] as num?)?.toDouble(),
      timestamp: json['timestamp'] as int,
    );
  }

  List<int> encodeToBytes() {
    return utf8.encode(jsonEncode(toJson()));
  }
}

File: aether_flow_apple/SharedInputEvent.swift
import Foundation

/// 共享輸入事件 (滑鼠/鍵盤)
public struct SharedInputEvent: Codable {
    public enum EventType: Int, Codable {
        case mouseMove = 0
        case mouseLeftDown = 1
        case mouseLeftUp = 2
        case mouseRightDown = 3
        case mouseRightUp = 4
        case keyDown = 5
        case keyUp = 6
        case scroll = 7
    }
    
    public let type: EventType
    public let x: Double? // 標準化座標 (0.0 - 1.0)
    public let y: Double?
    public let keyCode: Int?
    public let deltaX: Double?
    public let deltaY: Double?
    public let timestamp: TimeInterval
    
    public init(type: EventType, x: Double? = nil, y: Double? = nil, keyCode: Int? = nil, deltaX: Double? = nil, deltaY: Double? = nil) {
        self.type = type
        self.x = x
        self.y = y
        self.keyCode = keyCode
        self.deltaX = deltaX
        self.deltaY = deltaY
        self.timestamp = Date().timeIntervalSince1970
    }
}
